llvm/include/llvm/ADT/SmallVector.h

```diff
@@ -701,6 +701,10 @@ class SmallVectorImpl : public SmallVectorTemplateBase<T> {
     this->set_size(this->size() + NumInputs);
   }
 
+  template <typename RangeTy> void append_range(RangeTy &&R) {
+    this->append(R.begin(), R.end());
```

Comment:
Does this optimize around the temporaries here and move the elements? Should we have a test for this?

Reply:
I don't think this will move elements.  It should copy elements because we cannot reuse the storage used by the container that the given range is pointing to.

Reply:
We can't reuse the storage but we should be able to move the elements, no? E.g., skip allocations when appending a temporary range of `std::string`s. Maybe something based on `make_move_iterator` could work?

BTW, I think this should also use `adl_being`/`end`.

Reply:
I don't think this matches `std::vector` behaviour. A range being an rvalue shouldn't propagate to the results of its iterators.
See e.g. https://godbolt.org/z/dP4ohz9hM
This is problematic if you use a range that is some adaptor over an owning range as these are often RValues. Something like `vector.append_range(otherVector)` would have different behaviour than `vector.append_range(MutableArrayRef{otherVector})`

Reply:
Ah, good point.

Reply:
The `std::vector<std::vector<int>>` example will be nice to have in unittests/ADT/